### PR TITLE
fix(core): prevent panics from user-controlled pagination and string inputs

### LINKED ISSUE #2563: Bug: PageNumberPagination allows page_size=0 causing division by zero panic
## Bug Description

`PageNumberPagination::page_size` can be set to 0 via the builder. When `paginate()` is called, it performs `total_count / self.page_size` and `total_count % self.page_size`, both of which panic with division by zero. The same issue exists in `get_page()`. No validation prevents setting `page_size` to 0.

## Steps to Reproduce

1. Create `PageNumberPagination` with `page_size = 0`
2. Call `paginate()` or `get_page()`
3. Division by zero panic

## Expected Behavior

The builder should validate `page_size > 0` and reject zero values.

## Actual Behavior

Zero is accepted and causes a runtime panic.

## Affected Files

- `crates/reinhardt-core/src/pagination/page_number.rs:349-433`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #2564: Bug: ParamErrorContext::with_raw_value panics on multi-byte UTF-8 strings at byte boundary
## Bug Description

The truncation logic in `with_raw_value` uses byte-based slicing (`&value[..500]`) on a `String`. If the 500th byte falls in the middle of a multi-byte UTF-8 character (e.g., emoji, CJK characters), this panics with "byte index is not a char boundary."

## Steps to Reproduce

1. Create a `ParamErrorContext` with a raw value containing multi-byte UTF-8 characters where the 500th byte falls within a character
2. Call `with_raw_value()`
3. Panic: "byte index 500 is not a char boundary"

## Expected Behavior

Use `value.chars().take(n)` or `value.char_indices()` to truncate at a character boundary.

## Actual Behavior

Direct byte slicing on UTF-8 string causes panic.

## Affected Files

- `crates/reinhardt-core/src/exception/param_error.rs:88`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #2621: Bug: CursorPagination and RelayPagination panic on out-of-bounds cursor position when dataset shrinks
## Bug Description

Both `CursorPagination::paginate` and `RelayPagination::paginate` panic with an out-of-bounds slice index when a decoded cursor position exceeds the current item count. This occurs when items are removed from the dataset between paginated requests.

### Affected Locations

1. **CursorPagination** (`crates/reinhardt-core/src/pagination/cursor.rs:257-261`): When `start > total_count`, the slice `items[start..end]` panics because `start > end`.

2. **RelayPagination** (`crates/reinhardt-core/src/pagination/cursor/relay.rs:256, 267-271`): Same issue. Additionally, `decode(after_cursor)? + 1` can overflow if decoded position is `usize::MAX`.

## Steps to Reproduce

```rust
let paginator = CursorPagination::new().page_size(5);
let items: Vec<i32> = (0..10).collect();
let result = paginator.paginate(&items, None, "http://example.com/").unwrap();

// Simulate dataset shrinking (items deleted)
let smaller_items: Vec<i32> = (0..3).collect();
// Use cursor from first page (encoded position=5) against smaller dataset
// panic: slice index out of bounds
```

## Expected Behavior

When cursor position exceeds current dataset size, return an empty results page (not panic). The `start` value should be clamped to `total_count`.

## Actual Behavior

Panics with `slice index start is greater than end` or `index out of bounds`.

## Environment

- Crate: reinhardt-core
- Components: `pagination/cursor.rs`, `pagination/cursor/relay.rs`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

This PR prevents panic/crash vulnerabilities in reinhardt-core:

- Fix division by zero in `PageNumberPagination` with page_size=0 (#2563)
- Fix UTF-8 panic in `ParamErrorContext::with_raw_value` on multi-byte strings (#2564)
- Fix out-of-range cursor panic in `CursorPagination`/`RelayPagination` (#2621)

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)

## Motivation and Context

User-controlled inputs can cause server panics: zero page size triggers division by zero, multi-byte UTF-8 strings cause byte-index panics, and out-of-range cursor values cause slice panics. All three are denial-of-service vectors.

Fixes #2563, fixes #2564, fixes #2621

## How Was This Tested?

- 13 new tests covering edge cases (zero page size, Japanese/emoji strings, out-of-range cursors, empty datasets)
- All 1657 reinhardt-core tests pass
- `cargo make fmt-check` and `cargo make clippy-check` pass

## Checklist

- [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I have followed the [Commit Guidelines](../blob/main/instructions/COMMIT_GUIDELINE.md)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have formatted the code with `cargo make fmt-fix`
- [x] I have checked the code with `cargo make clippy-check`

### Type Label
- [x] `bug` - Bug fix

### Priority Label
- [x] `high` - Important fix or feature

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/reinhardt-core/src/pagination/cursor/relay.rs [hunk 1]
```diff
@@ -457,4 +462,74 @@ mod tests {
...
+			b"test-secret-key-for-unit-tests!!",
+		);
+		let paginator = RelayPagination::new().with_encoder(
+			crate::pagination::cursor::Base64CursorEncoder::with_secret_key(
+				b"test-secret-key-for-unit-tests!!",
+			),
+		);
+		let items: Vec<i32> = (1..=10).collect();
+		// Encode a cursor pointing beyond data length
+		let out_of_range_cursor = encoder.encode(100).unwrap();
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/src/pagination/cursor/relay.rs]: Fixed. Tests now create one `Base64CursorEncoder` instance, use it to encode the cursor, then move it into the paginator via `with_encoder()`. This ensures the same encoder state is used for both encoding and decoding.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
